File: 7-matrices/6-dp-and-linear-algebra-applications/main.py
# DP and Linear Algebra Applications - Python Example

def edit_distance(str1, str2):
    m, n = len(str1), len(str2)
    dp = [[0]*(n+1) for _ in range(m+1)]

    for i in range(m+1):
        dp[i] = i
    for j in range(n+1):
        dp[j] = j

    for i in range(1, m+1):
        for j in range(1, n+1):
            if str1[i-1] == str2[j-1]:
                dp[i][j] = dp[i-1][j-1]
            else:
                dp[i][j] = 1 + min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])
    return dp[m][n]

print("Edit distance between 'kitten' and 'sitting':", edit_distance("kitten", "sitting"))

File: 7-matrices/6-dp-and-linear-algebra-applications/main.cpp
// DP and Linear Algebra Applications - C++ Example

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

int editDistance(const std::string& str1, const std::string& str2) {
    int m = str1.size(), n = str2.size();
    std::vector<std::vector<int>> dp(m+1, std::vector<int>(n+1));

    for (int i = 0; i <= m; i++) dp[i][0] = i;
    for (int j = 0; j <= n; j++) dp[j] = j;

    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (str1[i-1] == str2[j-1]) dp[i][j] = dp[i-1][j-1];
            else dp[i][j] = 1 + std::min({dp[i-1][j], dp[i][j-1], dp[i-1][j-1]});
        }
    }
    return dp[m][n];
}

int main() {
    std::cout << "Edit distance between 'kitten' and 'sitting': " 
              << editDistance("kitten", "sitting") << std::endl;
    return 0;
}
